parse.y

```diff
@@ -123,25 +123,106 @@ hash_literal_key_p(VALUE k)
+        return 0;
+    }
+
+    /* Special case for __LINE__ and Integer */
+    if (OBJ_BUILTIN_TYPE(val) == T_NODE && nd_type(RNODE(val)) == NODE_LINE && RB_INTEGER_TYPE_P(lit)) {
+        return RNODE(val)->nd_loc.beg_pos.lineno != FIX2INT(lit);
+    }
+    if (OBJ_BUILTIN_TYPE(lit) == T_NODE && nd_type(RNODE(lit)) == NODE_LINE && RB_INTEGER_TYPE_P(val)) {
+        return RNODE(lit)->nd_loc.beg_pos.lineno != FIX2INT(val);
+    }
+
+    /* Special case for __FILE__ and String */
+    if (OBJ_BUILTIN_TYPE(val) == T_NODE && nd_type(RNODE(val)) == NODE_FILE && RB_TYPE_P(lit, T_STRING)) {
+        return rb_str_hash_cmp(rb_node_file_path_val(RNODE(val)), lit);
+    }
+    if (OBJ_BUILTIN_TYPE(lit) == T_NODE && nd_type(RNODE(lit)) == NODE_FILE && RB_TYPE_P(val, T_STRING)) {
+        return rb_str_hash_cmp(rb_node_file_path_val(RNODE(lit)), val);
+    }
+
+    if ((OBJ_BUILTIN_TYPE(val) == T_NODE) && (OBJ_BUILTIN_TYPE(lit) == T_NODE)) {
+        NODE *node_val = RNODE(val);
+        NODE *node_lit = RNODE(lit);
+        enum node_type type_val = nd_type(node_val);
+        enum node_type type_lit = nd_type(node_lit);
+
+        if (type_val != type_lit) {
+            return -1;
+        }
+        else if (type_lit == NODE_LINE) {
+            return node_val->nd_loc.beg_pos.lineno != node_lit->nd_loc.beg_pos.lineno;
+        }
+        else if (type_lit == NODE_FILE) {
+            return rb_parser_string_hash_cmp(RNODE_FILE(node_val)->path, RNODE_FILE(node_lit)->path);
+        }
+        else {
+            rb_bug("unexpected node: %s, %s", ruby_node_name(type_val), ruby_node_name(type_lit));
+        }
+    }
+    else if ((OBJ_BUILTIN_TYPE(val) == T_NODE) || (OBJ_BUILTIN_TYPE(lit) == T_NODE)) {
+        return -1;
+    }
+    else {
+        return rb_iseq_cdhash_cmp(val, lit);
+    }
+}
+
+static st_index_t
+node_cdhash_hash(VALUE a)
+{
+    switch (OBJ_BUILTIN_TYPE(a)) {
+      case T_NODE:{
+        enum node_type type = nd_type(RNODE(a));
+        switch (type) {
+          case NODE_LINE:
+            /* Same with Integer in rb_iseq_cdhash_hash */
+            return INT2FIX(RNODE(a)->nd_loc.beg_pos.lineno);
+          case NODE_FILE:
+            /* Same with String in rb_iseq_cdhash_hash */
+            return rb_str_hash(rb_node_file_path_val(RNODE(a)));
+          /* ruby_special_consts is greater than or equal to 0, then use -1, -2, -3, ... for other NODE */
+          default:
+            rb_bug("unexpected node: %s", ruby_node_name(type));
+        }
+      }
+      default:
+        return rb_iseq_cdhash_hash(a);
+    }
+}
+
 static int
 literal_cmp(VALUE val, VALUE lit)
 {
     if (val == lit) return 0;
     if (!hash_literal_key_p(val) || !hash_literal_key_p(lit)) return -1;
-    return rb_iseq_cdhash_cmp(val, lit);
+    return node_cdhash_cmp(val, lit);
 }
 
 static st_index_t
 literal_hash(VALUE a)
 {
     if (!hash_literal_key_p(a)) return (st_index_t)a;
-    return rb_iseq_cdhash_hash(a);
+    return node_cdhash_hash(a);
```

Comment:
Is it necessary to separate `node_cdhash_`-functions?

Reply:
Once all NODE_LIT is separated into dedecated NODE type, I will remove `rb_iseq_cdhash_hash` and `rb_iseq_cdhash_cmp` from parse.y. In such case, it might be good to define `node_cdhash_`-functions.
Am I misunderstanding something?